eZ/Publish/Core/Persistence/Legacy/Content/FieldHandler.php

```diff
@@ -447,6 +447,31 @@ public function deleteFields($contentId, VersionInfo $versionInfo)
     }
 
     /**
+     * Deletes translated fields and their external storage data for the given Content Versions.
+     *
+     * @param int $contentId
+     * @param \eZ\Publish\SPI\Persistence\Content\VersionInfo[] $versions
```

Comment:
Is this really necessary? Wouldn't it be better to get all the versions from `ContentHandler` inside of `deleteTranslationFromContentFields`?

Reply:
> Is this really necessary? Wouldn't it be better to get all the versions from `ContentHandler` inside of `deleteTranslationFromContentFields`? 
>
> AFAIU, if call: `deleteTranslationFromContentFields($contentId, [], $languageCode)` it will keep _zombie_ field data in storage.

It's necessary because custom implementation of FT External Storages require `VersionInfo` for `deleteFieldData`.

@Nattfarinn I saw two possible Service injections that could solve it the way you want:
- `Persistence\Legacy\Content\Handler`
- `Persistence\Legacy\Content\TreeHandler`

In both cases I got `ServiceCircularReferenceException` because both `Content Handler` and `TreeHandler` depend on `FieldHandler`. The obvious step would be to move this entire block which depends on `$versionInfo` to `TreeHandler`, but this is not very SOLID and/or DDD. The other idea would be to repeat simple logic which uses `$contentGateway` and `$contentMapper` to load `VersionInfo` but this is not DRY.

Reply:
Makes sense.